### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by juanraromero**

### CODE REVIEW

File: Assignment4/task07.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+ns = Namespace("http://somewhere#")
+
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+  print(s)
+################################################################
+print("SPARQL:")
+q1 = prepareQuery('''
+  SELECT ?subClass WHERE {
+    ?subClass rdfs:subClassOf ns:LivingThing.
+  }
```
Comment (anaigmo) [inline Assignment4/task07.py:44]: missing recursive subclassof relationships

**Comment by anaigmo** [comment]
The files do not follow the submission instructions, there should be inside a folder with your username and student id. I also left some comments to improve the assignment :)

File: Assignment4/juanraromero-23A248/task07.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+# TO DO
+# Visualize the results
+################################################################
+print("RDF:")
+from rdflib.plugins.sparql import prepareQuery
+
+ns = Namespace("http://somewhere#")
+
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task07.py:37]: this does not retireve subclasses of subclasses of living thing. Use a recursive function

File: Assignment4/juanraromero-23A248/task07.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  print(s)
+################################################################
+print("SPARQL:")
+q1 = prepareQuery('''
+  SELECT ?subClass WHERE {
+    ?subClass rdfs:subClassOf ns:LivingThing.
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task07.py:43]: same as above. Use `*`

File: Assignment4/juanraromero-23A248/task07.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+# Visualize the results
+
+ns = Namespace("http://somewhere#")
+
+################################################################
+print("RDF:")
+
+for s,p,o in g.triples((None, RDF.type, ns.Person)):
+    print(s)
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.Person)):
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task07.py:66]: same as above

File: Assignment4/juanraromero-23A248/task07.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+################################################################
+print("SPARQL:")
+
+q2 = prepareQuery('''
+  SELECT ?individual WHERE {
+    { ?individual rdf:type ns:Person
+    }
+    UNION
+    { ?individual rdf:type ?Class.
+    ?Class rdfs:subClassOf ns:Person.
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task07.py:78]: same as above

File: Assignment4/juanraromero-23A248/task07.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+################################################################
+print("RDF")
+for s,p,o in g.triples((None, FOAF.knows, ns.RockySmith)):
+  print(s)
+
+################################################################
+print("SPARQL")
+q4 = prepareQuery('''
+  SELECT ?person{
+    ?person foaf:knows ns:RockySmith.
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task07.py:142]: the exercise asks for a name, not a URI

File: Assignment4/juanraromero-23A248/task08.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+missing = []
+for (i,p) in allValues:
+  if (i,p) not in completed:
+    missing.append((i,p))
+
+for (i1,p1) in missing:
+  for (i2,p2,v) in values:
+    if i1 == i2 and p1 == p2:
+      g1.add((i1,p1,v))
+
```
Comment (dgarijo) [inline Assignment4/juanraromero-23A248/task08.py:68]: You are copying the KG in a variable, the operating with that variable. Why not use the KG directly?

**Review by dgarijo — CHANGES_REQUESTED** [review]
Minor changes requested

**Comment by anaigmo** [comment]
Submitted and mostly fixed, no more PRs allowed

**Pull request merged**